src/sync.h

```diff
@@ -276,7 +302,7 @@ using DebugLock = UniqueLock<typename std::remove_reference<typename std::remove
 //!
 //! The above is detectable at compile-time with the -Wreturn-local-addr flag in
 //! gcc and the -Wreturn-stack-address flag in clang, both enabled by default.
-#define WITH_LOCK(cs, code) [&]() -> decltype(auto) { LOCK(cs); code; }()
+#define WITH_LOCK(cs, code) (MaybeCheckNotHeld(cs), [&]() -> decltype(auto) { LOCK(cs); code; }())
```

Comment:
While playing with this I observed that it is possible to circumvent the thread safety warnings mechanism by wrapping the lock call in a lambda and calling this lambda immediately. That circumvention only works for mutexes that are a member of a class:

```cpp
Mutex mutex_global;

// Commented annotations would silence the warnings, clang 15
class C
{
public:
    Mutex mutex_member;

    void f1() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_member)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_member' [-Wthread-safety-analysis]
        LOCK(mutex_member);
    }

    void f2()
    {
        [&]() { LOCK(mutex_member); }(); // no need for annotations
    }

    void f3() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_global)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_global' [-Wthread-safety-analysis]
        LOCK(mutex_global);
    }

    void f4() // EXCLUSIVE_LOCKS_REQUIRED(!mutex_global)
    {
        // warning: calling function 'MaybeCheckNotHeld' requires negative capability '!mutex_global' [-Wthread-safety-analysis]
        [&]() /* EXCLUSIVE_LOCKS_REQUIRED(!mutex_global) */ { LOCK(mutex_global); }();
    }
};
```

Looks like the construct used for `WITH_LOCK()` in this PR relies on the `f2()` behavior from the example above. Is that behavior documented somewhere in the clang docs? If not, then it may be wiser to not depend on undocumented behavior.

Reply:
You can avoid the thread safety warnings in a lot of ways if you create aliases for either the locks or the variables being guarded; eg:

```c++
    int i GUARDED_BY(mutex_member) = 0;
    void f5()
    {
        int* i_alias = &i;
        *i_alias = 5;
    }
```

(Using `[&]` here is creating an implicit alias for all the member variables, though clang does a better job of analysing past the aliases there than in the above example...)

I think the best long term approach is to do away with any uses of both `GlobalMutex` and `RecursiveMutex`, at which point we can annotate the lambda in `WITH_LOCK` with `EXCLUSIVE_LOCKS_REQUIRED(!cs)` and we're done. Having it work in the meantime allows us to be more confident that we don't have any potential recursive cases when switching a `RecursiveMutex` to a regular `Mutex`, so seems like it aids that goal to me.

Reply:
Ok, this explains why `f2()` needs no annotations. I guess at some point the compiler may become smarter, detect the reference and issue a warning.

> I think the best long term approach is to do away with any uses of both `GlobalMutex` and `RecursiveMutex`, at which point we can annotate the lambda in `WITH_LOCK` with `EXCLUSIVE_LOCKS_REQUIRED(!cs)` and we're done.

I agree. Do you envision that we do not have mutexes in global scope, or we have ones that are of type `Mutex`?

Reply:
Anytime you currently have:

```c++
GlobalMutex g_mut;
int g_importantnum GUARDED_BY(g_mut) = 0;
std::string g_importantstr GUARDED_BY(g_mut);

void bump_importantint(int offset)
{
    LOCK(g_mut);
    g_importantnum += offset;
}
```

you can rewrite it as

```c++
class important_stuff
{
private:
    Mutex m_mut;
    int m_num GUARDED_BY(m_mut) = 0;
    std::string m_str GUARDED_BY(m_mut);
public:
    void bump(int offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mut);
};
important_stuff g_important;
```

or similar (depending on whether you can move all the code that uses the mutex into the class, or if you have to expose the mutex), which doesn't require very large changes to the code, but does make for a better fit with the thread safety analysis. Might also make it easy to later change `g_important` to not need to be a global too.

Reply:
> Do you envision that we do not have mutexes in global scope, or we have ones that are of type `Mutex`

I guess your reply above favors the "we do not have mutexes in global scope" case, right?